main.go

```diff
@@ -293,6 +354,104 @@ func runPeriodically(ctx context.Context, opts options, c *prometheus.CounterVec
 	}
 }
 
+type instantQueryRoundTripper struct {
+	r       http.RoundTripper
+	token   string
+	TraceID string
+}
+
+func newInstantQueryRoundTripper(token string, r http.RoundTripper) *instantQueryRoundTripper {
+	if r == nil {
+		r = http.DefaultTransport
+	}
+
+	return &instantQueryRoundTripper{
+		token: token,
+		r:     r,
+	}
+}
+
+func (r *instantQueryRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
+	if r.token != "" {
+		req.Header.Add("Authorization", "Bearer "+r.token)
+	}
+
+	resp, err := r.r.RoundTrip(req)
+	if err != nil {
+		return resp, err
+	}
+
+	r.TraceID = resp.Header.Get("X-Thanos-Trace-Id")
+
+	return resp, err
+}
+
+func query(
+	ctx context.Context,
+	l log.Logger,
+	endpoint *url.URL,
+	token string,
+	query querySpec,
+) (err error) {
+	level.Debug(l).Log("msg", "running specified query", "name", query.Name, "query", query.Query)
+
+	u := new(url.URL)
+	*u = *endpoint
+	u.Path = ""
+	r := newInstantQueryRoundTripper(token, nil)
+
+	c, err := promapi.NewClient(promapi.Config{
+		Address:      u.String(),
+		RoundTripper: r,
+	})
+	if err != nil {
+		err = fmt.Errorf("create new API client: %w", err)
+		return
+	}
+
+	a := promapiv1.NewAPI(c)
+
+	res, _, err := a.Query(ctx, query.Query, time.Now())
+	if err != nil {
+		err = fmt.Errorf("querying: %w", err)
+		return
+	}
+
+	level.Debug(l).Log("msg", "request finished", "name", query.Name, "response", res.String(), "trace-id", r.TraceID)
+
+	return nil
+}
+
+// doGetFallback will attempt to do the request as-is, and on a 405 it will fallback to a GET request.
+// Copied from the prometheus API client v1.2.1 (as it was removed afterwards).
+// https://github.com/prometheus/client_golang/blob/55450579111f95e3722cb93dec62fe9e847d6130/api/client.go#L64
```

Comment:
it is still there just moved in 
https://github.com/prometheus/client_golang/blob/057bfe813ae54e7b111200a9b1fecd04b6bf6be8/api/prometheus/v1/api.go#L927

Reply:
You can't access DoGetFallback as it's on an unexported object. I wanted to keep this PR smaller and make the read/write check use the "real" API client now as it's now possible to be replaced with this somewhat custom request.